Add microarray gene ID conversion example

cbethell opened the pull request:
## Purpose:
What issue(s) does your PR address?

This PR closes #113  (the issue on creating a microarray gene ID conversion example notebook).

## Strategy
What was your strategy for this new or edited analysis?

I followed the comments on issue #113 and found a new (larger than n = 4 samples) mouse microarray dataset.

The dataset used in this paper is a [mouse glioma cancer stem cell dataset](Cancer Stem Cells Are Enriched In The Side-Population Cells In A Mouse Model Of Glioma) with n = 15 samples (which has been uploaded to the S3 bucket for testing/review).

I then copied the analysis steps from the RNA-seq gene ID conversion notebook, but changed the annotation package being loaded in to that relevant to the mouse genome.

## Concerns/Questions for reviewers:
What things should reviewers look out for?

- Is the current workflow missing any vital steps?
- Do the steps under the `Explore mapped data frame` section seem helpful? Are the added steps clear? Is there anything that should be added/removed from this section?

## Analysis Pull Request Check List (roughly in order):

### Content checks
* [x] All `{{BLANKS}}` have been replaced with the correct content.
* [x] [Sources are cited](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#citing-sources-in-text)
* [x] Seed is set (if applicable)

### Formatting Checks
* [x] Removed any manual numbering of sections.
* [x] Removed any instances of chunk naming.
* [x] Spell checked any Rmd file or md file.
* [x] Comments and documentation are up to date.

### Add datasets to S3
* [x] Added [data and metadata files to S3](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#adding-datasets-to-the-S3-bucket).

### Docker/Snakemake rendering components
* [x] Added the `.html` [link to the navigation bar](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#add-new-analyses-to-the-navbar).
* [x] Any not yet added packages needed for this analysis have been added to the Dockerfile and it successfully builds.
* [x] In the Docker container, [snakemake was run for rendering](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#how-to-re-render-the-notebooks).

cansavvy commented on 02-microarray/gene-id-convert_microarray_01_ensembl.Rmd:
````diff
@@ -0,0 +1,272 @@
+    df
+  )
+```
+
````
I briefly mentioned this on the issue, but I think toward the goal of modeling "always look at your data" it may be good for us to include a few quick investigation steps that can tell a bit about "how the mapping went". This can include but isn't limited to answering: 
- How many gene identifiers are mapped versus not mapped?
- How many multi mappings does each gene identifier have?

cansavvy commented on 02-microarray/gene-id-convert_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,272 @@
+
+## Map Ensembl IDs to gene symbols
+
+This code chunk will return a message that indicates that there are many mappings between our keys and columns but this is okay and is what we expect since genes and transcripts do not have a 1:1 relationship.
```
I think we will need to hit this point harder and briefly mention the different options for `multiVals` argument, highlighting that different downstream analyses may benefit from particular strategies. Whatever strategy we end up using here, we should also use a different one when we update the rna-seq conversion module.

cansavvy reviewed (commented):
This is a good start!  This doesn't yet discuss the `multivals` argument that I mentioned here: https://github.com/AlexsLemonade/refinebio-examples/pull/212#discussion_r486332840 so the last check in you did isn't exactly reflective of what's going on with the mappings. 

I think the meaning of the `multivals` argument and how it influences the output is a worthwhile thing for us to dive into. Because if you are confused by this argument, our users will definitely be confused by this too (I don't like that they have a default for this, I think it makes it easy for people to miss it). We should aim for making the `multivals` argument and what it means, _as clear as possible_ for our users.

So I think next you may want to take another closer look at the `mapIds` docs and we can always schedule a zoom/Google hangout/Slack/Teams chat to discuss if that would be helpful.

cbethell commented:
> This is a good start! This doesn't yet discuss the `multivals` argument that I mentioned here: [#212 (comment)](https://github.com/AlexsLemonade/refinebio-examples/pull/212#discussion_r486332840) so the last check in you did isn't exactly reflective of what's going on with the mappings.
> 
> I think the meaning of the `multivals` argument and how it influences the output is a worthwhile thing for us to dive into. Because if you are confused by this argument, our users will definitely be confused by this too (I don't like that they have a default for this, I think it makes it easy for people to miss it). We should aim for making the `multivals` argument and what it means, _as clear as possible_ for our users.
> 
> So I think next you may want to take another closer look at the `mapIds` docs and we can always schedule a zoom/Google hangout/Slack/Teams chat to discuss if that would be helpful.

Per a discussion with @cansavvy re the above comment, the plan moving forward is as follows:
- Supply the `multiVals` argument with the `"list"` option
- Use `reshape2::melt()` to get that large list into a more manageable data frame
- Explore the melted data frame using some of the steps already used in this PR (to look at multi mappings, etc.)
- Then rerun the `mapIds()` function, this time supplying the `multiVals` argument with the`"filter"` option which will remove all elements with multiple matches

cbethell commented:
@cansavvy, I _believe_ that I implemented the plan that we discussed and outlined in the [last comment of this PR](https://github.com/AlexsLemonade/refinebio-examples/pull/212#issuecomment-692251653).

I also implemented the suggestions you provided through your most recent review of this PR (the main ones being changing the file name, keeping the use of `IDs` consistent throughout the notebook, and some sentence restructuring).

Please let me know if I misinterpreted any of your suggestions or missed any!

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,368 @@
+From here you can customize this analysis example to fit your own scientific questions and preferences. 
+
+refine.bio data comes with gene level data with Ensembl IDs. 
+Although this example script uses Ensembl IDs from Mouse, (<i>Mus musculus</i>), to obtain gene symbols this script can be easily converted for use with different species or annotation types <i>eg</i> protein IDs, gene ontology, accession numbers.
```
```suggestion
Although this example notebook uses Ensembl IDs from Mouse, (<i>Mus musculus</i>), to obtain gene symbols this notebook can be easily converted for use with different species or annotation types <i>eg</i> protein IDs, gene ontology, accession numbers.
```

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,368 @@
+Although this example script uses Ensembl IDs from Mouse, (<i>Mus musculus</i>), to obtain gene symbols this script can be easily converted for use with different species or annotation types <i>eg</i> protein IDs, gene ontology, accession numbers.
+
+<b>For different species</b>, wherever the abbreviation `org.Mm.eg.db` or `Mm` is written, it must be replaced with the respective species abbreviation <i>eg</i> for<i> Homo sapiens</i> `org.Hs.eg.db` or `Hs` would be used.
+In the case of our [RNA-seq gene identifier annotation example notebook](https://github.com/AlexsLemonade/refinebio-examples/blob/master/03-rnaseq/gene-id-convert_rnaseq_01_ensembl.Rmd), a Zebrafish (<i>Danio rerio</i>) dataset is used, meaning `org.Dr.eg.db` or `Dr` would also need to be used there.
```
I know its' not ready yet, but we should probably have this go to the html instead of the Rmd.

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,368 @@
+
+See our Getting Started page with [instructions for package installation](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#what-you-need-to-install) for a list of the other software you will need, as well as more tips and resources.
+
+In this analysis, we will be using the `org.Mm.eg.db` R package [@Carlson2019].
```
```suggestion
In this analysis, we will be using the `org.Mm.eg.db` R package [@Carlson2019].
[Other species can be used](#using-a-different-refinebio-dataset-with-this-analysis). 
```

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,368 @@
+
+## Map Ensembl IDs to gene symbols
+
+This code chunk will return a message that indicates that there are many mappings between our keys and columns but this is okay and is what we expect in most cases since genes and transcripts do not have a 1:1 relationship.
```
This message will change depending on the `multiVals` argument, so we either need to reframe this sentence or delete it. If we do keep it, we should put it after the user would see this message.

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,368 @@
+## Map Ensembl IDs to gene symbols
+
+This code chunk will return a message that indicates that there are many mappings between our keys and columns but this is okay and is what we expect in most cases since genes and transcripts do not have a 1:1 relationship.
+The `mapIds()` function has a `multiVals` argument which denotes what to do when there are multiple mapped values for a single gene identifier.
```
I like this section! Great!

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,368 @@
+
+It looks like we have gene symbols that were successfully mapped to the Ensembl IDs we provided.
+However, the data is now in a `list` object, making it a little more difficult to explore.
+We are going to turn our list object into a data frame object in the next chunk.
```
Nice! I like this too!

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
````diff
@@ -0,0 +1,368 @@
+```{r}
+# Let's make our object a bit more manageable for exploration by turning it into a data frame using the `reshape2::melt()` function
+mapped_df <- mapped_list %>%
+  reshape2::melt()
````
I *think* you can use the `value` arugment to name what you'd like the column to be called, but you'd have to try that. I think we should either do that or use `rename()` so its clear what these columns are.

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
````diff
@@ -0,0 +1,368 @@
+```{r}
+multi_mapped <- mapped_df %>%
+  # Let's group by the Ensembl ID's in the `L1` column
+  dplyr::group_by(L1) %>%
````
Yeah, we might want to rename `L1` before we get to this step.

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,368 @@
+
+Looks like we have some cases where 3 gene symbols mapped to a single Ensembl ID.
+We have a total of 130 out of 17984 Ensembl IDs with multiple mappings to gene symbols.
+We are not too worried about the 130 IDs with multiple mappings, so we will filter them out for the purpose of having 1:1 mappings for our downstream analysis.
```
```suggestion
If we are not too worried about the 130 IDs with multiple mappings, we can filter them out for the purpose of having 1:1 mappings for our downstream analysis.
```

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,368 @@
+
+## Map Ensembl IDs to gene symbols -- filtering out multi mappings
+
+The next code chunk will rerun the `mapIds()` function, this time supplying the `"filter"` option to the `multiVals` argument.
```
```suggestion
The next code chunk we will rerun the `mapIds()` function, this time supplying the `"filter"` option to the `multiVals` argument.
```

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,368 @@
+  # Turn our list into a data frame
+  reshape2::melt() %>%
+  # Now let's give the `value` column a name that is more relevant to the values it holds
+  dplyr::rename("Symbol" = "value")
```
Yes, this bit, can we do this above?

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,368 @@
+  multiVals = "filter"
+) %>%
+  # Turn our list into a data frame
+  reshape2::melt() %>%
```
I might be wrong, but I don't think this melt is necessary, I *think* you should be able to put this straight into a data frame since it should be the same length as the original data.

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
````diff
@@ -0,0 +1,368 @@
+  dplyr::rename("Symbol" = "value")
+```
+
+Let's copy the same code we used above to see if we have any multi mappings in our new data frame.
````
I don't like copying the code. I also don't think this step is necessary. We can trust `mapIds()` will work as advertised. Plus this notebook has gotten a tad long. I would go straight to writing to TSV and call it a day.

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,368 @@
+# Resources for further learning
+
+- Marc Carlson has prepared a nice [Introduction to Bioconductor Annotation Packages](https://bioconductor.org/packages/release/bioc/vignettes/AnnotationDbi/inst/doc/IntroToAnnotationPackages.pdf) [@Carlson2020]
+- [More on Annotation mapping functions](https://www.bioconductor.org/packages/release/bioc/vignettes/AnnotationFuncs/inst/doc/AnnotationFuncsUserguide.pdf) [@Edwards2020]
```
I wouldn't share this alternative annotation package's user guide, I find it more confusing than helpful. Easier just to share AnnotationDbi user guide tutorials as opposed to sending our users to something else that they would have to learn from scratch.

cbethell replied:
Gotcha, that makes sense. I removed this reference but haven't found any additional user guide tutorials thus far, do you have any suggestions or personal favorites for this particular case?

cansavvy replied:
I don't think I have anything else. We can add more if we find something.

cansavvy reviewed (commented):
This looks great! Just have a few comments about a couple things to rearrange, I think this is close!

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -297,47 +298,27 @@ head(multi_mapped)
 ## Map Ensembl IDs to gene symbols -- filtering out multi mappings
 
-The next code chunk will rerun the `mapIds()` function, this time supplying the `"filter"` option to the `multiVals` argument.
+The next code we chunk we will rerun the `mapIds()` function, this time supplying the `"filter"` option to the `multiVals` argument.
```
```suggestion
The next code chunk we will rerun the `mapIds()` function, this time supplying the `"filter"` option to the `multiVals` argument.
```

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -297,47 +298,27 @@ head(multi_mapped)
-  dplyr::arrange(desc(gene_symbol_count)) %>%
-  # Filter to include only the rows with multi mappings
-  dplyr::filter(gene_symbol_count > 1)
+filtered_mapped_df <- data.frame(
```
Should probably save `data` with it too.

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -297,47 +298,27 @@ head(multi_mapped)
-  dplyr::arrange(desc(gene_symbol_count)) %>%
-  # Filter to include only the rows with multi mappings
-  dplyr::filter(gene_symbol_count > 1)
+filtered_mapped_df <- data.frame(
```
```suggestion
filtered_mapped_df <- data.frame(
     df, 
```
This isn't formatted properly, but you get the idea.

cbethell replied:
@cansavvy, I get an error here due to the difference in length between `df` (nrows = 17918) and the results in `filtered_mapped_df` (nrows = 17854). Do you think we should pull the genes out of the rownames and do an `inner_join` here?

cansavvy replied:
Oh of course. Forgot about that. Will need an inner_join. I can take a look after add that if you want.

cbethell replied:
Okay @cansavvy, you can take a look at the `inner_join` step and let me know if I need to make any changes 👍

cansavvy reviewed (approved):
Two tiny-ish comments! LGTM after that!

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,352 @@
+
+## Explore gene ID conversion
+
+Now, let's take a look at our mapped object to see how the mapping went.
```
```suggestion
Now, let's take a look at our `mapped_list` to see how the mapping went.
```

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,352 @@
+    keys = df$Gene,
+    column = "SYMBOL", # Replace with the type of gene identifiers you would like to map to
+    keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
+    multiVals = "filter"
```
```suggestion
    multiVals = "filter" # This will drop any `Gene`s that have multiple matches
```

cansavvy reviewed (approved):
Just a couple little comments. Looks like there's some template changes that might need to be done too; some that got lost: https://github.com/AlexsLemonade/refinebio-examples/pull/216#discussion_r490194300
But I will probably file a separate PR to change these everywhere since this will affect every module and I haven't yet figured out the full extent of what needs to be changed.

cansavvy commented on 02-microarray/gene-id-annotation_microarray_01_ensembl.Rmd:
```diff
@@ -0,0 +1,352 @@
+
+# Purpose of this analysis
+
+The purpose of this notebook is to provide an example of mapping gene IDs for microarray data obtained from refine.bio using `AnnotationDbi` packages.
```
Probably should have a citation here too.

jaclyn-taroni reviewed (approved):
LGTM

jaclyn-taroni commented:
I'm playing with the settings that require branches to be up to date before merging. I am going to merge this despite the lack of status check (that I do not currently understand).
